layabout.py

```diff
@@ -1,3 +1,287 @@
+            interval: The number of seconds to wait between fetching events
+                from the Slack API.
+            retries: The number of retry attempts to make if a connection to
+                Slack if not established or is lost.
+            backoff: The strategy used to determine how long to wait between
+                retries. Must take as input the number of the current retry and
+                output a :obj:`float`. If absent an exponential backoff
+                strategy will be used.
+            until: The condition used to evaluate whether this method
+                terminates. Must take as input an :obj:`list` of :obj:`dict`
+                representing Slack RTM API events and return a :obj:`bool`. If
+                absent this method will run forever.
+
+        Raises:
+            FailedConnection: If connecting to the Slack API fails.
+        """
+        backoff = backoff or _exponential
+        until = until or _forever
+
+        # The initial connection may use a given token or attempt to use an
+        # environment variable.
+        if not self.connect(token=token):
+            if not self._reconnect(retries=retries, backoff=backoff):
+                raise FailedConnection('Failed to connect to the Slack API')
+
+        # TODO: Should we force callers to handle KeyboardInterrupt on their
+        # own, or should we try to handle it for them? 🤔
+        while True:
+            try:
+                # Fetch new RTM events from the API.
+                events = self.slack.rtm_read()
+
+            # This is necessary to handle an error caused by a bug in Slack's
+            # Python client. For more information see
+            # https://github.com/slackhq/python-slackclient/issues/127
+            #
+            # TODO: The TimeoutError could be more elegantly resolved by making
+            # a PR to the websocket-client library and letting them coerce that
+            # exception to a WebSocketTimeoutException.
+            except (WebSocketConnectionClosedException, TimeoutError):
+                log.debug('Lost connection to the Slack API, attempting to '
+                          'reconnect')
+                # Attempt to re-use our existing SlackClient and token.
+                if self._reconnect(retries=retries, backoff=backoff):
+                    log.debug('Reconnected to the Slack API')
+                    # We continue so we can fetch new events before proceeding.
+                    continue
+
+                raise FailedConnection('Failed to reconnect to the Slack API')
+
+            if not until(events):
+                # TODO: Is this even a good debugging message?
+                log.debug('Terminal condition met')
+                break
+
+            # Handle events!
+            for event in events:
+                type_ = event.get('type')
+                # TODO: Should * handlers be run first?
```

Comment:
I don't know if they should be handled first but there should definitively be an ordering of some kind (instead of the source code ordering which is going to be first handle call first).

My gut says that * should be handled last since they're most likely to fail?

Reply:
Are they the most likely to fail? I don't see any reason why they should be.

When you say a definitive ordering do you mean there should be a priority ordering that's user-configurable? Something like a CSS `z-index`?

By
> ... first handle call first ...

I assume you're calling attention to the fact that the order of handler registration determines the order in which handlers are called. If that's explicitly documented (which it's currently not) what's the problem with that?

Reply:
> Are they the most likely to fail?

Having to handle all of the different kinds of input in a single function seems more complex to me. Though on second thought hopefully people aren't writing handlers that will stack trace.

> When you say a definitive ordering do you mean there should be a priority ordering that's user-configurable? Something like a CSS z-index?

Probably overkill. Potentially I could see adding a sorting callback for the user to add their own ordering.

> If that's explicitly documented (which it's currently not) what's the problem with that?

That's a totally fine answer to me.

Reply:
I think having handler registration order be call order is pretty reasonable once it's documented.  Then if the user desires, they can specify the call order without any further explicit support in `Layabout`.

And if for some reason they _need_ their function definitions laid out in a different order than they want the handler call order, they can just not use the decorator and call `handle` explicitly in the order they need.

It seems to me the added complexity of providing an explicit ordering mechanism is not worthwhile.  Not without a good use case certainly.

I don't really have an opinion on whether `*` handlers should go first or last.  I think either choice is going to be problematic for certain uses.  That might be an argument for having `pre-*` and `post-*` handlers, but I don't think it's necessary right now.  Again, I would want a good motivating use case first.